tools/buildgen/extract_metadata_from_bazel_xml.py

```diff
@@ -247,119 +198,193 @@ def _external_dep_name_from_bazel_dependency(bazel_dep):
         return None
 
 
-def _expand_intermediate_deps(target_dict, public_dep_names, bazel_rules):
-    # Some of the libraries defined by bazel won't be exposed in build.yaml
-    # We call these "intermediate" dependencies. This method expands
-    # the intermediate deps for given target (populates library's
-    # headers, sources and dicts as if the intermediate dependency never existed)
-
-    # use this dictionary to translate from bazel labels to dep names
-    bazel_label_to_dep_name = {}
-    for dep_name in public_dep_names:
-        bazel_label_to_dep_name[_get_bazel_label(dep_name)] = dep_name
+def _compute_transitive_metadata(
+        rule_name: str, bazel_rules: Any,
+        bazel_label_to_dep_name: Dict[str, str]) -> None:
+    """Computes the final build metadata for Bazel target with rule_name.
+
+    The dependencies that will appear on the deps list are:
+
+    * Public build targets including binaries and tests;
+    * External targets, like absl, re2.
+
+    All other intermediate dependencies will be merged, which means their
+    source file, headers, etc. will be collected into one build target. This
+    step of processing will greatly reduce the complexity of the generated
+    build specifications for other build systems, like CMake, Make, setuptools.
+
+    The final build metadata are:
+    * _TRANSITIVE_DEPS: all the transitive dependencies including intermediate
+                        targets;
+    * _COLLAPSED_DEPS:  dependencies that fits our requirement above, and it
+                        will remove duplicated items and produce the shortest
+                        possible dependency list in alphabetical order;
+    * _COLLAPSED_SRCS:  the merged source files;
+    * _COLLAPSED_PUBLIC_HEADERS: the merged public headers;
+    * _COLLAPSED_HEADERS: the merged non-public headers;
+    * _EXCLUDE_DEPS: intermediate targets to exclude when performing collapsing
+      of sources and dependencies. 
+
+    For the collapsed_deps, the algorithm improved cases like:
+
+    The result in the past:
+        end2end_tests -> [grpc_test_util, grpc, gpr, address_sorting, upb]
+        grpc_test_util -> [grpc, gpr, address_sorting, upb, ...]
+        grpc -> [gpr, address_sorting, upb, ...]
+    
+    The result of the algorithm:
+        end2end_tests -> [grpc_test_util]
+        grpc_test_util -> [grpc]
+        grpc -> [gpr, address_sorting, upb, ...]
+    """
+    bazel_rule = bazel_rules[rule_name]
+    direct_deps = _extract_deps(bazel_rule, bazel_rules)
+    transitive_deps = set()
```

Comment:
I don't think so.
1. when I made this change I regenerated the yaml files and there was no difference in the generated files
2. the transitive deps aren't actually exposed anywhere in the resulting yaml file, it's just an internal field (fields like collapsed_deps get sorted, which would cancel any effect of transitive deps being ordered).
3. it doesn't matter since the algorithm is agressively eliminating transitive public dependencies in the output. If something depends on both grpc and gpr, only `grpc` will end up in the final list of dependencies.  So (as you pointed out in previous dicussion), the build order doesn't matter since in all cases where it would matter (e.g. gpr coming before grpc, which would be incorrect), the final collapsed_deps will  never contain both libraries that have a dependency relationship between them.